src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -2,45 +2,102 @@
         {
-            //.NET Native doesn't support AssemblyLoadContext.GetLoadContext. 
-            // Initializing the _loadContext in the .ctor would cause
-            // .NET Native builds to fail because the .ctor is called. 
-            // However, LoadFromPathImpl is never called in .NET Native, so 
-            // we do a lazy initialization here to make .NET Native builds happy.
-            if (_loadContext == null)
+            if (!ShouldLoad(fullPath))
             {
-                AssemblyLoadContext loadContext = AssemblyLoadContext.GetLoadContext(typeof(DefaultAnalyzerAssemblyLoader).GetTypeInfo().Assembly);
+                throw new InvalidOperationException();
+            }
+
+            AssemblyLoadContext? loadContext;
 
-                if (System.Threading.Interlocked.CompareExchange(ref _loadContext, loadContext, null) == null)
+            var fullDirectoryPath = Path.GetDirectoryName(fullPath) ?? throw new ArgumentException(message: null, paramName: nameof(fullPath));
+            lock (_guard)
+            {
+                if (!_loadContextByDirectory.TryGetValue(fullDirectoryPath, out loadContext))
                 {
-                    _loadContext.Resolving += (context, name) =>
-                    {
-                        Debug.Assert(ReferenceEquals(context, _loadContext));
-                        return Load(name.FullName);
-                    };
+                    loadContext = new DirectoryLoadContext(fullDirectoryPath, this);
+                    _loadContextByDirectory[fullDirectoryPath] = loadContext;
                 }
             }
 
-            return LoadImpl(fullPath);
+            // We allow analyzer loaders to "group" assemblies by directory,
+            // and then perform a substitution before the assembly is actually loaded.
+            var pathToLoad = GetPathToLoad(fullPath);
+            return loadContext.LoadFromAssemblyPath(pathToLoad);
         }
 
-        protected virtual Assembly LoadImpl(string fullPath) => _loadContext.LoadFromAssemblyPath(fullPath);
+        private sealed class DirectoryLoadContext : AssemblyLoadContext
+        {
+            private readonly string _directory;
+            private readonly DefaultAnalyzerAssemblyLoader _loader;
+
+            public DirectoryLoadContext(string directory, DefaultAnalyzerAssemblyLoader loader)
+            {
+                _directory = directory;
+                _loader = loader;
+            }
+
+            protected override Assembly? Load(AssemblyName assemblyName)
+            {
+                // Any compilers assembly or any assembly referenced by the compilers needs to "win"
+                // over a user-specified version of that assembly.
+                var currentAssembly = Assembly.GetExecutingAssembly();
+                if (assemblyName.Name is "Microsoft.CodeAnalysis" or "Microsoft.CodeAnalysis.CSharp" or "Microsoft.CodeAnalysis.VisualBasic"
```

Comment:
There are more than just the 3 of these.

Example:
- Microsoft.CodeAnalysis.Common
- Microsoft.CodeAnalysis.FSharp (if it actually exists)

Reply:
IIRC Microsoft.CodeAnalysis.Common is just the NuGet package which contains Microsoft.CodeAnalysis.dll.

I'm definitely open to better methods of getting the set of assembly names to "delegate to the compiler's context", but strictly it only seems necessary to do this trick for assemblies that the C# and VB compilations depend on. Are we sure that "assemblies whose name starts with `Microsoft.CodeAnalysis`" is not going to catch something we didn't intend to include?

For example, if an analyzer somehow referenced `Microsoft.CodeAnalysis.Workspaces.dll`, it wouldn't necessarily be our job here to ensure that it's the same Workspaces assembly loaded in the context containing the compiler assemblies. I could be wrong about that, though.

Reply:
I think we have addressed the problem of how to identify and specially handle compiler assemblies so will resolve this conversation.